### ISSUE TITLE
is_flat_shape() fails in show_dense_layer()

### ISSUE THREAD

**Issue by swilson314** [Dobiasd/frugally-deep#89]
This model has a number of errors when calling convert_model().

[model.zip](https://github.com/Dobiasd/frugally-deep/files/2536442/model.zip)

I'm using tf-embedded Keras, so that may account for some of the problems, e.g. not being able to set the model name.

The other problem source seems due to the fact that I am using a 2d conv layer where one of the axis is time_steps=10. Also, because I'm using temporal weighting, I have to specify batch_size=10 via batch_input_shape.

Many of the errors I've been able to work around per below.

The problem I haven't been able to work around occurs in show_dense_layer() with a layer that has input_shape=(10,864), so is_flat_shape() fails.

```
	# SBW 2018.10.31 model.model doesn't exist for some tf models.
        # if model.model == model:
        if True:
```

```
	# SBW 2018.10.31 Can't set name.
        # set_model_name(model, name)
```
```
    def set_shape_idx_0_to_1(shape):
        """Change first element in tuple to 1."""
        # SBW 2018.10.31 Doesn't work when batch_size is set via batch_input_shape.
        if shape[0] is not None:
            return shape
        shape_lst = list(shape)
        shape_lst[0] = 1
        shape = tuple(shape_lst)
        return shape
```

**Comment by swilson314** [comment]
If I comment the is_flat_shape() assert, the program runs to completion and outputs model.json. When I try to load the model from c++ I get a "Invalid inputs shape" exception during the testing of the model.

**Comment by Dobiasd** [comment]
Currently frugally-deep only supports dense layers with 1-dimensional input tensors, i.e., with a flatten layer in front of it.

But if I understand correctly, you don't want this, but the other case described [here](https://stackoverflow.com/a/43237727/1866775) instead.

So this would be something that has to be implemented. Are you interested in giving it a try in a PR? Otherwise I could have a look at it too.

**Comment by Dobiasd** [comment]
Good finding! Up to now frugally-deep did only supports dense layers with 1-dimensional input tensors, i.e., with a flatten layer in front of it.

But if I understand correctly, you don't want this, but the other case described [here](https://stackoverflow.com/a/43237727/1866775) instead.

So I just [implemented support for it](https://github.com/Dobiasd/frugally-deep/commit/2d9cefee6e120c95ebd2ba064ccc061c7302fcbf). Please checkout the latest version and re-open the issue if this does not solve your problem. :)

**Comment by swilson314** [comment]
Fantastic, many thanks! I'll check out the fix.

Just to be clear, there is a flatten layer just before the dense layer, I believe the issue is that the batch_size is specified:

```
	mlp_model.add(Conv2D(filters1, (2, inputs_per_fold), strides=(1, inputs_per_fold), padding='valid', activation='tanh', batch_input_shape=batch_input_shape))
	mlp_model.add(MaxPooling2D(2))
	mlp_model.add(Flatten())
	mlp_model.add(Dense(hidden_final, activation='tanh'))
	mlp_model.add(Dense(time_steps, activation='tanh'))
	mlp_model.add(Reshape((time_steps, 1)))

```

**Comment by Dobiasd** [comment]
This should already have worked without any problem with the old version! Now **that** does trigger my interest! :D

So I tried to reproduce your model in a minimal example:

```python
# save_model.py
from keras.layers import Dense, Conv2D, Flatten, MaxPooling2D, Reshape
from keras.models import Sequential

filters1 = 8
inputs_per_fold = 3
batch_input_shape = (None, 64, 64, 3)
hidden_final = 20
time_steps = 10

def build_model():
    mlp_model = Sequential()

    mlp_model.add(Conv2D(filters1, (2, inputs_per_fold),
        strides=(1, inputs_per_fold), padding='valid',
        activation='tanh', batch_input_shape=batch_input_shape))
    mlp_model.add(MaxPooling2D(2))
    mlp_model.add(Flatten())
    mlp_model.add(Dense(hidden_final, activation='tanh'))
    mlp_model.add(Dense(time_steps, activation='tanh'))
    mlp_model.add(Reshape((time_steps, 1)))

    mlp_model.compile(loss='categorical_crossentropy', optimizer='nadam')
    return mlp_model

def main():
    model = build_model()
    model.summary()
    model.save('keras_model.h5', include_optimizer=False)

if __name__ == '__main__':
    main()
```

```cpp
// main.cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_model.json");
}
```

```bash
python3 save_model.py
python3 convert_model.py keras_model.h5 fdeep_model.json
g++ -std=c++14 main.cpp -o main
./main
```

```
Loading json ... done. elapsed time: 0.016220 s
Running test 1 of 1 ... done. elapsed time: 0.006761 s
Loading, constructing, testing of fdeep_model.json took 0.053682 s overall.
```

Does this minimal example mimic your model correctly? If so, can you please run it on your machine with the non-updated frugally-deep version, that produces the problem, and tell me about the results?

I'd really love to understand what is going on.

**Comment by swilson314** [comment]
Thanks for much for your prompt responses!!

You need the following changes to your code above to model this problem. The setting of batch_size in batch_input_shape is required by compile() because I'm using sample_weights.

```
batch_input_shape = (10, 64, 64, 3)
mlp_model.compile(optimizer=opt, loss='mse', metrics=None, sample_weight_mode='temporal')
```

**Comment by Dobiasd** [comment]
Oh, OK. I did not yet consider models like this. Converting it should fails at multiple points, for example [here (`layer.input_shape[0] is None`)](https://github.com/Dobiasd/frugally-deep/blob/92561d33e6fb802db59391c454cca2212e7506cd/keras_export/convert_model.py#L200).

I'll see what I can do. :+1:

**Comment by swilson314** [comment]
Excellent, thanks. I'm trying a workaround where I rebuild the model without using sample_weights, load the original weights, and then save. I'll let you know.

That said, any work you do on this will pay off when support for LSTM is added since that's another case where batch_size!=None.

**Comment by Dobiasd** [comment]
I tried to support it with some quick modifications, but it seems to be not that simple.

From what I see, using `sample_weight_mode='temporal'` changes the inner behavior of certain layers, e.g., the number/shape of output tensors differs.

I guess this is something I have to first understand more deeply to be able to implement. So despite you helping me by having brought in the idea to support dense layers with non-flattened input (even though this was not your actual intention), I can not help you right now with support of your actual model in frugally-deep.

Let's see how the [LSTM stuff](https://github.com/Dobiasd/frugally-deep/pull/83) turns out regarding this.

---

If it is a viable option for you to not use `sample_weight_mode='temporal'`, this would certainly make your model quite standard, and thus supported out-of-the box by frugally-deep. :)

**Comment by swilson314** [comment]
The workaround appears to have solved the problem. Thus, I still train with 'temporal', but then, before frugally-deep, I build a version of the model (using previously saved weights) without 'temporal'.

Again, thanks so much for your help with this! I'll make additions to the code base where I can :)
